File: CodingPractice/src/main/java/Arrays/Sorting/mergeSort.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Sorting;

/**
 *
 * @author Hp
 */
public class mergeSort 
{
    public static void main(String arg[])
    {
        int[] arr = {14,20,5,9,10, 30, 20};
        mergeSort merge = new mergeSort();
        
        merge.mergeSort(arr,0, arr.length - 1);
        merge.print(arr);
        
    }    
    
    public void print(int[] arr)
    {
        for(int i = 0; i< arr.length ; i++)
        {
            System.out.println(arr[i]);
        }
    }
    public void mergeSort(int[] arr, int low, int high)
    {
        
        if(low < high)
        {
            int mid = (low+high)/2;            
            mergeSort(arr, low, mid);
            mergeSort(arr, mid+1, high);            
            merge(arr, low, mid, high);
        }              
        
    }
    
    public void merge(int[] nums, int low, int mid, int high)
    {
        int len1= mid - low +1;
        int len2 = high - mid;
        
        int[] len_one = new int[len1];
        int[] len_two = new int[len2];
        
        for(int i = 0; i< len1; i++)
        {
            len_one[i] = nums[low+i];
        }
        for(int j = 0; j <len2 ; j++)
        {
            len_two[j] = nums[mid+1+j];
        }
        int x = 0,y = 0,k = low;
        while(x < len1 && y < len2)
        {
            if(len_one[x] <= len_two[y])
            {
                nums[k] = len_one[x];
                x++;
                k++;
            }
            else
            {
                nums[k] = len_two[y];
                y++;
                k++;
            }
        }
        
        while(x < len1)
        {
            nums[k] = len_one[x];
            x++;
            k++;
        }
        while(y <len2)
        {
            nums[k] = len_two[y];
            y++;k++;
        }
    }
    
}

File: Solved in Python/LeetCode/Sorting/mergeSort.py
# class MergeSort(object):
    # def main(self):
    #     array = [6,3,2,5,7,9,1]
    #     print(self.mergeSort(array, 0, len(array) - 1))
import sys
    
sys.setrecursionlimit(2500)
def mergeSort(arr, low, high):

    if low < high:
        middle =(low + high )//2
        mergeSort(arr, low, middle)
        mergeSort(arr, middle +1, high)
        merge(arr, low, middle, high)

    

def merge(arr, low, middle, high):
    l1 = middle - low + 1
    l2 = high - middle
    len_one = [0] * l1
    len_two = [0] * l2
    for i in range(0, l1):
        len_one[i] = arr[low + i]

    for j in range(0, l2):
        len_two[j] = arr[middle+1+j]

    x = 0
    y = 0
    k = low
    while x < l1 and y < l2:
        if len_one[x] <= len_two[y]:
            arr[k] = len_one[x]
            x += 1
            k += 1

        else:
            arr[k] = len_two[y]
            y += 1
            k += 1

    while x < l1:
        arr[k] = len_one[x]
        x += 1
        k += 1

    while y < l2:
        arr[k] = len_two[y]
        y += 1
        k += 1

    
array = [6,3,2,5,7,9,1]
    
mergeSort(array, 0, len(array) - 1)
print(array)

